storybookjs/storybook
# Svelte: remount when resetting args in controls

Closes #21470



## What I did

This PR makes the Svelte renderer force remount the story when a user resets all args in the controls panel.

It does this by listening for the `RESET_STORY_ARGS` event, saves the information that a story is being reset, and uses that information in `renderToCanvas` to remount instead of just updating props.

There's a possible race condition here, because the `RESET_STORY_ARGS` event will both trigger this custom listener and also trigger `renderToCanvas`. If for some reason `renderToCanvas` reacts to the event **before** the custom listener, then it won't know that it should remount, and it will most likely remount on the following render, eg. triggered by a regular arg change.
I've been unable to produce this out-of-order triggering so for now it's purely theoretical.



There's still an issue here though, that when any singular control gets unset, that will not cause a full remount and the component will not revert to default but rather act as if the value was `undefined`. From a purely Svelte-rendering perspective this is expected, the same thing happens if your Svelte components do that (set a prop to something and then sets it to `undefined`). But maybe from a Storybook user perspective this is unexpected, at least I would expect it to go back to the default.
This scenario is only valid for the color control as it is the only control that can be unset. But I suspect we'll make this a universal feature in the future.

## How to test

1. See new test story being successful: https://630873996e4e3557791c069c-eagyhguotu.chromatic.com/?path=/story/stories-renderers-svelte-args--remount-on-reset-story-args
2. Try it out with [the default Button story](https://630873996e4e3557791c069c-eagyhguotu.chromatic.com/?path=/story/example-button--primary):
    1. Open the Button story in a Svelte project
    2. Change the "size" control to "large"
    3. Reset args
    4. See that it goes back to the default
    5. Do the same thing but [on the `next` branch](https://630873996e4e3557791c069c-ucioaavosy.chromatic.com/?path=/story/example-button--primary) and see that the button loses its size completely because `size` becomes `undefined`. 



## Checklist



      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

#### Maintainers

      make sure to add the `ci:merged` or `ci:daily` GH label to it.

`["cleanup", "BREAKING CHANGE", "feature request", "bug", "documentation", "maintenance", "dependencies", "other"]`
